contracts/evoting/evoting.go

```diff
@@ -601,3 +616,64 @@ func (e evotingCommand) cancelElection(snap store.Snapshot, step execution.Step)
 
 	return nil
 }
+
+// verifyIdentity is a utility function to verify if a public key is associated
+// to a member of the roster or not. Returns no error if it's the case.
+func verifyIdentity(roster authority.Authority, publicKey []byte) error {
+	pubKeyIterator := roster.PublicKeyIterator()
+	isAMember := false
+
+	for pubKeyIterator.HasNext() {
+		key, err := pubKeyIterator.GetNext().MarshalBinary()
+		if err != nil {
+			return xerrors.Errorf("failed to serialize a public key from the roster : %v ", err)
+		}
+
+		if bytes.Equal(publicKey, key) {
+			isAMember = true
+		}
+	}
+
+	if !isAMember {
+		return xerrors.Errorf("public key not associated to a member of the roster: %x", publicKey)
+	}
+
+	return nil
+}
+
+// SemiRandomStream implements cipher.Stream
+type SemiRandomStream struct {
+	// Seed is the seed on which should be based our random number generation
+	seed []byte
+
+	stream *rand.Rand
+}
+
+func NewSemiRandomStream(seed []byte) (SemiRandomStream, error) {
+	if len(seed) > 8 {
+		seed = seed[0:8]
```

Comment:
I'm a little bit concerned that we use only the first 8 bytes. It means that one only need to forge the 8 first bytes of the SHA256 to compromise the random source.

Reply:
Maybe then using the hash of the hash would do the trick.

Reply:
Forget it, the hash does not depend on the prover since it comes from the shuffled ballots + the electionID